build/android/src/main/java/net.minetest.minetest/AssetCopyTask.java

```diff
@@ -0,0 +1,289 @@
+		if (!tempFolder.exists()) {
+			if (!tempFolder.mkdir()) {
+				Log.e(TAG, "Failed to tmp folder " +
+					tempFolder.getAbsolutePath());
+			}
+		} else {
+			for (File f : tempFolder.listFiles()) {
+				Log.v(TAG, "Deleting: " + f.getAbsolutePath());
+				if (!f.delete()) {
+					Log.e(TAG, "Failed to delete " + f.getAbsolutePath());
+				}
+			}
+		}
+
+		// Build lists from prepared data
+		buildFolderList();
+		buildFileList();
+
+		// Scan file list
+		processFileList();
+
+		copyStarted = true;
+
+		for (int i = 0; i < toCopy.size(); ++i) {
+			try {
+				publishProgress(i);
+				copyFile(toCopy.get(i));
+			} catch (IOException e) {
+				Log.e(TAG, "Copying file: " +
+					toCopy.get(i) + " failed.", e);
+			}
+		}
+
+		activity.startActivity(new Intent(activity, GameActivity.class));
+
+		return "";
+	}
+
+	private void migrateBasePath() {
+		File oldBasePath = new File(Environment.getExternalStorageDirectory(), "Minetest");
+		if (!oldBasePath.exists())
+			return;
+		if (oldBasePath.renameTo(basePath)) {
+			Log.i(TAG, "Migrated old base path.");
+			return;
+		}
+		Log.e(TAG, "Failed to migrate old base path from " +
+			oldBasePath.getAbsolutePath() + " to " +
+			basePath.getAbsolutePath() + " deleting it instead.");
+		if (!oldBasePath.delete()) {
+			Log.e(TAG, "Failed to delete old base path, leaving it.");
+		}
+	}
+
+	private void copyFile(String filename) throws IOException {
+		boolean asset_size_unknown = assetSizeUnknown.contains(filename);
+		long file_size = -1;
+
+		if (asset_size_unknown) {
+			File test = new File(basePath, filename);
+			if (test.exists())
+				file_size = test.length();
+		}
+
+		InputStream src;
+		try {
+			src = activity.getAssets().open(gameAssetPrefix + filename);
+		} catch (IOException e) {
+			Log.e(TAG, "Failed to copy file (not in assets): " +
+				filename, e);
+			return;
+		}
+
+		long asset_size;
+
+		if (asset_size_unknown) {
+			asset_size = getFullSize(src);
+			if (asset_size == file_size) {
+				src.close();
+				return;
+			}
+			src.reset();
+		}
+
+		OutputStream dst;
+		File dstFile = new File(basePath, filename);
+		try {
+			dst = new FileOutputStream(dstFile);
+		} catch (IOException e) {
+			Log.e(TAG, "Failed to copy file (couldn't open output file): " +
+				dstFile.getAbsolutePath(), e);
+			src.close();
+			return;
+		}
+
+		byte[] buf = new byte[8192];
+		int len;
+		long total_filesize = 0;
+		while ((len = src.read(buf)) > 0) {
+			dst.write(buf, 0, len);
+			total_filesize += len;
+		}
+
+		dst.close();
+		src.close();
+		Log.v(TAG, "Copied file: " + filename +
+			" (" + total_filesize + " bytes).");
+	}
+
+	/// Update progress bar
+	@Override
+	protected void onProgressUpdate(Integer... progress) {
+		if (copyStarted) {
+			String filename = toCopy.get(progress[0]);
+			int text_res = R.string.copying;
+			if (assetSizeUnknown.contains(filename))
+				text_res = R.string.checking;
+
+			activity.filenameView.setText(String.format(
```

Comment:
don't access activity members like this, the Java way would be to:

1. Make ProgressUpdateReceiver Interface
2. Inherit the activity from it
3. Change setActivity to setReceiver()
4. Call onProgressUpdate() from here

If this is too Java-y, then at least do all this in a function in the activity. (add a onProgressUpdate function to the activity, call that directly rather than through an interface). The interface's purpose is to stop this class having a direct concrete reference to the activity and so dissuade abuse